### DISCUSSION TITLE
EMQX as "message bus" for scaling LDAP

### DISCUSSION THREAD

**Discussion by nikoladsp** [emqx/emqx#8273] [Q&A] [answered]
Hi,
I am thinking is it possible to have an architecture as given [here](https://ibb.co/d53ZFfK) using EMQX?

What I would like to achieve is to have scaled OpenLDAP system; where requests between master LDAP nodes can be load-balanced (say using round-robin), and each master node can have any number of replicas (zero or more). Master and backup nodes are "near", say in same local network (or same host, but different docker images), while masters can be in different physical locations (different cities). Master-Backup uses LDAP default replication, however Master-Master does not use LDAP replication, but rather custom protocol which would sit on top of EMQX. This way I can change "type" of Master-Master replication depending on client needs.

I am not sure are there any better solutions or if this what I want is just plain stupid. Same question applies for other types of DBs, like any key-value DB (Redis).

Side question: what is the recommended JavaScript client?

Many hanks in advance.

Kindest regards

**Comment by zmstone** [comment, marked as answer]
Hi.
I’m afraid MQTT protocol in general is not quite suitable for database transaction/commit log replication.
Kafka is likely a better choice at the moment.
This recommendation is based on the assumptions below:

* Replica may restart and fetch the missing window during its down time (and the down time can be arbitrary long)
* Master may send data to a queue before the replica is ready (queue needs to be declared or created before a replica is up and running)
* Replica may need to replay messages sent before it “subscribes” to a replication queue. I.e., replay. (eg., to bootstrap a new replica from a snapshot)

**Reply by nikoladsp** [reply]
Thank you very much!

As for EMQX, I plan to use it as message broker in the system, so my question for recommended JavaScript library still stands: is Paho OK or something else is better choice?

And one more, if you don't mind:
Say I have an EMQX broker instance receive 10 messages and while not all messages have been published, it goes down due some failure After it is up again, will all remaining unpublished messages be published?

Best regards

**Reply by zmstone** [reply]
Hi @nikoladsp 

For js client, below are the links for your reference 
https://github.com/mqttjs/MQTT.js library 
https://github.com/emqx/MQTTX a desktop app 
http://tools.emqx.io/about a web app

In MQTT only messages published to a subscriber’s persistent session are persisted.

In order to survive service restart:
* EMQX enterprise edition offers message persistence to external database.
* Community edition 5.1 will offer on disk message persistence.

**Reply by nikoladsp** [reply]
Many thanks!
